### ISSUE TITLE
False Positive Error for not Enqueuing Styles for String Value

### ISSUE THREAD

**Issue by cedon** [WordPress-Coding-Standards/WordPress-Coding-Standards#1360]
PHP: v7.2.4
PHPCS: v3.0.2
WPCS: v0.14.0

I am getting the following error from WPCS when running it on my `functions.php` file:
```
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 223 | ERROR | Stylesheets must be registered/enqueued via
     |       | wp_enqueue_style
----------------------------------------------------------------------
```

The line in question is actually a string being used in a filter to modify enqueued CSS files, specifically allowing me to add additional strings like `integrity` and `crossorigin` since WordPress currently does not have that functionality on its own.

The function in question is:

```
**
 * Filters certain styles in theme_scripts() and adds 'integrity', 'crossorigin', or other attributes if they are
 * in an array defined within the function.
 *
 * @param $string $html The HTML element of the enqueued style.
 * @param $string $handle The handle used to register the style.
 * @param $string $href The stylesheet source URL.
 * @param $string $media The media attribute of the style sheet.
 *
 * @return string The HTML element of the enqueued style as-is or with additional attributes added.
 */
function theme_style_integrity( $html, $handle, $href, $media ) {
	$styles = array(
		'fontawesome-cdn' => 'integrity="sha384-oi8o31xSQq8S0RpBcb4FaLB8LJi9AT8oIdmS1QldR8Ui7KUQjNAnDlJjp55Ba8FG" crossorigin="anonymous"',
	);

	if ( array_key_exists( $handle, $styles ) ) {
		// This $html variable declaration causes the error
                $html = '<link rel="stylesheet" src="' . $href . '" ' . $styles[ $handle ] . ' type="text/css">';
	}
	return $html;
}
add_filter( 'style_loader_tag', 'theme_style_integrity', 10, 4 );
```

**Comment by jrfnl** [comment]
IMHO this is not a false positive.

You are "misusing" the filter and disregarding that anyone else could be using that same `$handle` and calling the filter too.

By overwriting the `<link>` tag this way, you could be throwing away the changes made by another plugin/theme.

**Comment by cedon** [comment]
How is it not a false positive? It literally is telling me "You are not using `wp_enqueue_style()` to load this stylesheet" when I am declaring the string value of a variable that is to be passed out of the function.

As far as why I am doing this, you're right, I shouldn't have to. Support for this should have been added to WordPress long ago. There is a 3 year old ticket (https://core.trac.wordpress.org/ticket/33948) on this issue. But until that feature is added, I have to do it this way because a) I don't want to run a plugin, and b) I want to secure all scripts that are not being served directly by me. I want the added protection of the browser not executing scripts instead of my visitors' computers being hijacked to mine cryptocurrency. 

As far as using the same handle, the developer documentation for `wp_enqueue_script()` and `wp_enqueue_style()` state that `$handle` should be unique so a plugin should be using handles that incorporate the slug and if not, I can always adjust mine as well. But either way only the $handle values I deliberately put in that array get messed with. Everything else should be passed on through as is. If there is a way to specifically target scripts from my theme's enqueue function I'd be happy to add that.

**Comment by jrfnl** [comment]
> How is it not a false positive? It literally is telling me "You are not using wp_enqueue_style() to load this stylesheet" when I am declaring the string value of a variable that is to be passed out of the function to the result of wp_enqueue_style().

@cedon It's not a false positive because you still have a hard-coded `<link>` tag in your code, so you are overruling the enqueue you did earlier effectively (no matter that it's for a valid reason).

You could change your code to be non-conflicting with other plugins and themes though as AFAICS, you are only adding a property to the link tag. What about something like this:

```php
if ( array_key_exists( $handle, $styles ) ) {
	// Add additional properties.
	$html = str_replace ( '>', ' ' . $styles[ $handle ] . '>', $html );
}
```

Or maybe even better, have a look at the example here on how to add attributes to the stylesheet generation: https://codex.wordpress.org/Class_Reference/WP_Styles#Other_Attributes

**Comment by jrfnl** [comment]
@cedon Glad to hear you got it working.

> So WPCS cannot see the context of the `return $html` statement?

Code can be written in lots of different ways and PHPCS (including WPCS) can't easily look across files, so if you have the `add_filter()` call in another part of the code or for instance, within a class `__construct()` method, WPCS would not be able to tell that that particular function is hooked into the `style_loader_tag` filter.

**Comment by GaryJones** [comment]
> I tried using $wp_styles->add_data() but apparently that is restricted to certain values for $key as they did **not** show up when I modified the function to this

Between https://core.trac.wordpress.org/ticket/37162, https://core.trac.wordpress.org/ticket/33948 and https://core.trac.wordpress.org/ticket/22249, you're right that this isn't yet supported.

Still, the string replacement that doesn't include adding in `<link ` can be used for now, without triggering the CS violation.
